### ISSUE TITLE
Wrongly suggesting missing implementations of a subclassed dataclass protocol.

### ISSUE THREAD

**Issue by sksg** [microsoft/pyright#10386]
**Describe the bug**
The issue is that pyright wrongly-in my opinion-shows an error for missing implementations of a subclass of a dataclass protocol i.e. a `typing.Protocol` with the `dataclasses.dataclass` decorator. However, as can be seen by running the example in python, indeed subclassing such a dataclass protocol works as exepcted. I believe the inheritance of a protocol should be, in effect, the same as inheriting a normal base class. In any case, this is how it behaves. See also the use of non-protocols works as expected.

**Code or Screenshots**
See minimal working example here:
[https://github.com/sksg/pyright-dataclass-protocol-issue](https://github.com/sksg/pyright-dataclass-protocol-issue)

The code is also included here:

```python
from dataclasses import dataclass
from typing import Protocol


@dataclass(kw_only=True)
class dataclass_one_base:
    """A base class that exposes a dataclass interface."""

    a: int
    b: str

    def method(self) -> None:
        """A method that does something."""
        print(f"Method called with a={self.a}, b={self.b}")


@dataclass(kw_only=True)
class dataclass_one_protocol(Protocol):
    """A protocol that exposes a dataclass interface."""

    a: int
    b: str

    def method(self) -> None:
        """A method that does something."""
        print(f"Method called with a={self.a}, b={self.b}")


@dataclass(kw_only=True)
class dataclass_one(dataclass_one_base):
    """A class that implements the dataclass_one_protocol."""

    pass


@dataclass(kw_only=True)
class dataclass_one_prime(dataclass_one_protocol):
    """A class that implements the dataclass_one_protocol."""

    pass


# An inherited dataclass base types correctly.

instance_one = dataclass_one(a=1, b="example one")
instance_one.method()  # This works

# However, an inherited dataclass protocol does not type correctly although it works as expected.

instance_one = dataclass_one_prime(a=1, b="example one")
instance_one.method()  # This works

# Interestingly, if we create the types dynamically, they both type correctly.
dataclass_one_dynamic = type("dataclass_one_dynamic", (dataclass_one_base,), {})
instance_one_dynamic = dataclass_one_dynamic(a=1, b="example")

# Yet the dynamic protocol version does not actually work as expected, even if it types correctly.
dataclass_one_dynamic_prime = type(
    "dataclass_one_dynamic_prime", (dataclass_one_protocol,), {}
)

try:
    instance_one_dynamic_prime = dataclass_one_dynamic_prime(a=1, b="example")
    # There is no type error here. At runtime, however, it raises:
    # TypeError: dataclass_one_dynamic.__init__() takes exactly one argument (the instance to initialize)
except TypeError as e:
    print(f"Expected TypeError occurred: {e}")

# We can make the dynamic version work by using the dataclass decorator, as it most likely makes a new `__init__`
# method.
dataclass_one_dynamic_prime_prime = dataclass(
    type("dataclass_one_dynamic_prime_prime", (dataclass_one_protocol,), {})
)
instance_one_dynamic_prime_prime = dataclass_one_dynamic_prime_prime(a=1, b="example")


@dataclass(kw_only=True)
class dataclass_two_base:
    """Another base class that exposes a dataclass interface."""

    x: float
    y: str

    def another_method(self) -> None:
        """Another method that does something."""
        print(f"Another method called with x={self.x}, y={self.y}")


@dataclass(kw_only=True)
class dataclass_two_protocol(Protocol):
    """Another protocol that exposes a dataclass interface."""

    x: float
    y: str

    def another_method(self) -> None:
        """Another method that does something."""
        print(f"Another method called with x={self.x}, y={self.y}")


@dataclass(kw_only=True)
class combined_dataclass(dataclass_one_base, dataclass_two_base):
    """A class that combines two protocols."""

    pass


@dataclass(kw_only=True)
class combined_dataclass_prime(dataclass_one_protocol, dataclass_two_protocol):
    """A class that combines two protocols."""

    pass


# Multiple inherited base dataclasses types correctly.

instance_combined = combined_dataclass(a=2, b="example two", x=2.0, y="test")
instance_combined.method()  # This works
instance_combined.another_method()


# Yet, again the combined dataclass protocol does not type correctly

instance_combined_prime = combined_dataclass_prime(
    a=2, b="example two", x=2.0, y="test"
)
instance_combined_prime.method()  # This still works
instance_combined_prime.another_method()

```

To show the issue run

```bash
pyright ./example_script.py
```

in your shell which in my shell outputs

> \src\pyright-dataclass-protocol-issue\example_script.py
>   \src\pyright-dataclass-protocol-issue\example_script.py:50:16 - error: Cannot instantiate abstract class "dataclass_one_prime"
>     "dataclass_one_protocol.a" is not implemented
>     "dataclass_one_protocol.b" is not implemented (reportAbstractUsage)
>   \src\pyright-dataclass-protocol-issue\example_script.py:124:27 - error: Cannot instantiate abstract class "combined_dataclass_prime"
>     "dataclass_two_protocol.x" is not implemented
>     "dataclass_two_protocol.y" is not implemented
>     and 2 more... (reportAbstractUsage)
> 2 errors, 0 warnings, 0 informations

You can, however, run the example script and see that indeed they do work as
expected at runtime

```bash
python ./example_script.py
```

and the output is

> Method called with a=1, b=example one
> Method called with a=1, b=example one
> Expected TypeError occurred: dataclass_one_dynamic_prime.__init__() takes exactly one argument (the instance to initialize)
> Method called with a=2, b=example two
> Another method called with x=2.0, y=test
> Method called with a=2, b=example two
> Another method called with x=2.0, y=test

Notice that the dynamic creation of the dataclasses do not use the base class
`__init__` functions correctly. This might be an effect of the protocols
explicitly disallowing direct instantiation.

**VS Code extension or command-line**
Pyright version 1.1.400 with CLI.

**Comment by erictraut** [comment]
The link in your example doesn't work for me. I get a 404.

Could you provide a minimal, self-contained example in code form within the bug report itself?

**Comment by sksg** [comment]
Link fixed and example code added.

**Comment by erictraut** [comment]
What you're doing here is very atypical, and it's not clear to me why you're doing this. You're using `@dataclass` in conjunction with a structural type (protocol) definition.

The [typing spec](https://typing.python.org/en/latest/spec/index.html) provides little or no guidance about how these constructs interact. I don't think the authors of the PEPs that introduced `dataclass` and `Protocol` to the type system intended for these to be used together — or at least never considered what the intended behavior would be if they were. I'd argue that it's ambiguous.

Normally, an attribute definition in a `Protocol` is implicitly considered "abstract" if it is not supplied with a default value. In this case, one could perhaps argue that the behind-the-scenes "magic" of the `dataclass` mechanism is supplying values to these attributes, so these attributes shouldn't be considered abstract. I think that's the assumption you're making here.

I'll note that pyright's current behavior in this case matches that of mypy, which was the reference implementation for both `dataclass` and `Protocol`. For this reason, I think pyright's current behavior here is defensible. If you'd like to advocate for a different behavior, you could propose a clarification in the Python typing spec. Pyright attempts to [conform with the typing spec](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html), so if you can get consensus on a clarification, I would then change the behavior to match the spec. [Here](https://github.com/python/typing-council#decisions) is the process by which you can go about this.

Short of changing the typing spec, my recommendation is to avoid using `Protocol` and `dataclass` together in the same class definition. This avoids the ambiguity altogether. Through multiple inheritance, you can still inherit from a dataclass base class _and_ a protocol. I don't see any benefit to combining the two.

If your intent is to define a structural definition for some aspects of a dataclass, you should do so explicitly rather than relying on the synthesized signatures of methods that are synthesized by the `@dataclass` mechanism. For example, if you want to define a protocol that requires the implementation of the `__eq__` method (which is synthesized by `@dataclass`), then you should add an `__eq__` method explicitly to your protocol definition.

I'm going to close this issue because I don't plan to take any action on it unless/until the intended behavior is clarified in the typing spec.

**Comment by erictraut** [comment]
More generally, I wonder if the typing spec should be changed to explicitly disallow any class decorators on a Protocol class definition (other than, perhaps, `@typing.final`). It's not clear to me that supporting class decorators for protocols is desirable, since the intent is ambiguous.

**Comment by sksg** [comment]
Thanks, Eric, for always speedy feedback.

I see your points and will look into proposing a clarification for this type of use.

For the record, I needed to define multiple `dataclass` mixin types e.g.

```python

@dataclass
class mixin_1:
    a: int
    b: str

@dataclass
class mixin_2:
    x: bool
    y: float

@dataclass
class concrete(mixin_1, mixin_2):  # This is a common method for mixing dataclasses
    pass

```

And I needed protocols with the same interfaces. These are necessary for some function signatures

```python
func(instance: mixin_2):  # This now only supports nominal subtyping rather than structural subtyping 
    ...
```

Protocols are explicitly made to support default implementations of methods, and subclassing a protocol is the standard way to inherit these default implementations. I do believe it is very reasonable to expect something like this to work (which is does) and to type-check.

However, I agree that the combination of `Protocol`s and synthesized methods like, for example, from the `dataclass`  decorator have likely not been considered properly. In particular, I agree that

> When a decorator is applied to a class, it's not clear whether the resulting (post-decorated) class should be considered a protocol.

And this indeed seems to be the best argument against the behavior I am expecting.
